consensus/spos/worker_test.go

```diff
@@ -582,6 +584,80 @@ func TestWorker_ProcessReceivedMessageRedundancyNodeShouldResetInactivityIfNeede
 	assert.True(t, wasCalled)
 }
 
+func TestWorker_ProcessReceivedMessageEquivalentMessageShouldReturnError(t *testing.T) {
+	t.Parallel()
+
+	workerArgs := createDefaultWorkerArgs(&statusHandlerMock.AppStatusHandlerStub{})
+	wrk, _ := spos.NewWorker(workerArgs)
+
+	equivalentBlockHeaderHash := workerArgs.Hasher.Compute("equivalent block header hash")
+	pubKey := []byte(wrk.ConsensusState().ConsensusGroup()[0])
+	headerBytes := make([]byte, 100)
+	_, _ = rand.Read(headerBytes)
+
+	bodyBytes := make([]byte, 100)
+	_, _ = rand.Read(bodyBytes)
+
+	cnsMsg := consensus.NewConsensusMessage(
+		equivalentBlockHeaderHash,
+		nil,
+		nil,
+		nil,
+		pubKey,
+		bytes.Repeat([]byte("a"), SignatureSize),
+		int(bls.MtBlockHeaderFinalInfo),
+		0,
+		chainID,
+		[]byte("01"),
+		signature,
+		signature,
+		currentPid,
+		nil,
+	)
+	buff, _ := wrk.Marshalizer().Marshal(cnsMsg)
+
+	cnsMsgEquiv := consensus.NewConsensusMessage(
+		equivalentBlockHeaderHash,
+		nil,
+		nil,
+		nil,
+		pubKey,
+		bytes.Repeat([]byte("b"), SignatureSize),
+		int(bls.MtBlockHeaderFinalInfo),
+		0,
+		chainID,
+		[]byte("01"),
+		signature,
+		signature,
+		currentPid,
+		nil,
+	)
+	buffEquiv, _ := wrk.Marshalizer().Marshal(cnsMsgEquiv)
+
+	err := wrk.ProcessReceivedMessage(
+		&p2pmocks.P2PMessageMock{
+			DataField:      buff,
+			PeerField:      currentPid,
+			SignatureField: []byte("signature"),
+		},
+		fromConnectedPeerId,
+		&p2pmocks.MessengerStub{},
+	)
+	assert.NoError(t, err)
+
+	equivMsgFrom := core.PeerID("from other peer id")
+	err = wrk.ProcessReceivedMessage(
+		&p2pmocks.P2PMessageMock{
+			DataField:      buffEquiv,
+			PeerField:      currentPid,
+			SignatureField: []byte("signatureEquiv"),
+		},
+		equivMsgFrom,
+		&p2pmocks.MessengerStub{},
+	)
+	assert.Equal(t, spos.ErrEquivalentMessageAlreadyReceived, err)
```

Comment:
test also the inner counter? Or a special unit test for the reset & read of the counters?

Reply:
improved the test